### PR TITLE
[APB-1973][HO,MP] upgrade to new auth

### PR DESCRIPTION
**Pull request by habeeb01**

### CODE REVIEW

File: app/uk/gov/hmrc/agentsubscription/GuiceModule.scala [hunk 1]
```diff
@@ -38,8 +38,8 @@ class GuiceModule(environment: Environment, configuration: Configuration) extend
     bind(classOf[HttpPost]).toInstance(WSHttp)
     bind(classOf[HttpGet]).toInstance(WSHttp)
     bind(classOf[DesConnector])
-    bind(classOf[AuthConnector])
     bind(classOf[AuditConnector]).toInstance(MicroserviceGlobal.auditConnector)
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentsubscription/GuiceModule.scala]: May we avoid having a global object here and use DI?

File: app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala [hunk 2]
```diff
@@ -16,40 +16,88 @@
...
+import uk.gov.hmrc.auth.core.{Enrolment, _}
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.microservice.controller.BaseController
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+case class Provider(providerId: String, providerType: String)
+
+class AuthConnector @Inject()(metrics: Metrics, microserviceAuthConnector: MicroserviceAuthConnector)
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala]: We should add `@Singleton` here

File: app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala [hunk 3]
```diff
@@ -16,40 +16,88 @@
...
-    val absoluteUserDetailsUrl = new URL(authorityUrl, userDetailsLink).toString
-    httpGet.GET[UserDetails](absoluteUserDetailsUrl)
+  private val AuthProvider: AuthProviders = AuthProviders(GovernmentGateway)
+  private val agentEnrol = "HMRC-AS-AGENT"
+  private val agentEnrolId = "AgentReferenceNumber"
+  private val isAnAgent = true
+
+  implicit val hc: HeaderCarrier = new HeaderCarrier
+
+  def forSubscription(action: SubscriptionAuthAction) = Action.async(parse.json) {
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala]: IMHO this is not a good idea to mix connector stuff with auth actions, better to keep them separate as before, WDYT?

File: app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala [hunk 4]
```diff
@@ -16,40 +16,88 @@
...
+        case e @ _ => {
+          Logger.error("Failed to auth", e)
+          GenericUnauthorized
+        }
+      }
   }
 
-  def enrolments(authority: Authority)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[List[Enrolment]] = {
-    httpGet.GET[List[Enrolment]](authority.absoluteEnrolmentsUrl)
+  def forRegistration(action: RegistrationAuthAction) = Action.async {
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala]: the same comment about coupling connector with actions applies here

File: app/uk/gov/hmrc/agentsubscription/controllers/RegistrationController.scala [hunk 5]
```diff
@@ -18,24 +18,28 @@ package uk.gov.hmrc.agentsubscription.controllers
...
 
 @Singleton
-class RegistrationController @Inject()(service: RegistrationService, override val authConnector: AuthConnector)
-  extends BaseController with AuthActions {
+class RegistrationController @Inject()(service: RegistrationService)
+                                      (implicit metrics: Metrics, microserviceAuthConnector: MicroserviceAuthConnector)
+  extends AuthConnector(metrics, microserviceAuthConnector) with BaseController {
 
-  private[controllers] def getRegistrationBlock(utr: Utr, postcode: String): RequestWithAuthority[AnyContent] => Future[Result] = { implicit request =>
+  private[controllers] def getRegistrationBlock(utr: Utr, postcode: String)
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentsubscription/controllers/RegistrationController.scala]: I think we can find a better name than `getRegistrationBlock` or simply inline this function if is applied only once?

File: app/uk/gov/hmrc/agentsubscription/controllers/RegistrationController.scala [hunk 6]
```diff
@@ -44,14 +48,19 @@ class RegistrationController @Inject()(service: RegistrationService, override va
...
+
   }
 
   private def badRequest(code: String) = {
     Future successful BadRequest(Json.obj("code" -> code))
   }
 
-  def getRegistration(utr: Utr, postcode: String): Action[AnyContent] =
-    withAgentAffinityGroup.async(getRegistrationBlock(utr, postcode))
+  def getRegistration(utr: Utr, postcode: String) = forRegistration {
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentsubscription/controllers/RegistrationController.scala]: I find names like `forRegistration`a little bit too abstract, does it reveals what it does without forcing a reader to look deep inside the code?

File: app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala [hunk 7]
```diff
@@ -16,40 +16,88 @@
...
+
+  implicit val hc: HeaderCarrier = new HeaderCarrier
+
+  def forSubscription(action: SubscriptionAuthAction) = Action.async(parse.json) {
+    implicit request =>
+      authorised(AuthProvider).retrieve(affinityGroupAllEnrols) {
+        case Some(affinityG) ~ allEnrols ⇒
+          (isAgent(affinityG), extractEnrolmentData(allEnrols.enrolments, agentEnrol, agentEnrolId)) match {
+            case (`isAnAgent`, Some(_)) => action(request)
+            case (`isAnAgent`, None) => action(request)
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala]: those 2 cases seem to be the same?

Comment (arturopala) [reply app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala]: if an agent is not enrolled should we proceed?

File: app/uk/gov/hmrc/agentsubscription/GuiceModule.scala (hunk 1 above)
Comment (habeeb01) [reply app/uk/gov/hmrc/agentsubscription/GuiceModule.scala]: addressed

File: app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala (hunk 2 above)
Comment (habeeb01) [reply app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala]: addressed

File: app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala (hunk 7 above)
Comment (habeeb01) [reply app/uk/gov/hmrc/agentsubscription/connectors/AuthConnector.scala]: both cases have been merged

File: app/uk/gov/hmrc/agentsubscription/controllers/RegistrationController.scala (hunk 6 above)
Comment (habeeb01) [reply app/uk/gov/hmrc/agentsubscription/controllers/RegistrationController.scala]: renamed "forRegistration" to "affinityGroupAndCredentials"
renamed "forSubscription" to "affinityGroupAndEnrolments"

**Pull request merged**